colexechash: clean up the hash table a bit

yuzefovich opened the pull request:
This PR contains several commits that refactor the hash table a bit with the main goal of not accessing `hashTableProbeBuffer.distinct` slice (now named `foundNull`) from the hash aggregator. This is a nice-to-have cleanup for the vectorized hash group joiner. Last commit optimizes the code a bit. The benchmarks of `colexec` [package](https://gist.github.com/yuzefovich/98c4bd6766eb9ee95bc3b13949905c85) show mixed results, but mostly positive improvements (ignore benches unrelated to hash aggregator and unordered distinct).

Epic: None

Release note: None

yuzefovich committed:
colexechash: some mechanical changes to the hash table

This commit moves `FindBuckets` method into the templated code and inlines
`FindNext` method. This is a purely mechanical change in preparation for the
follow-up commit that will be templating the touched code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101204)
<!-- Reviewable:end -->

yuzefovich committed:
colexechash: rename hashTableProbeBuffer.distinct to make it clearer

This commit renames `hashTableProbeBuffer.distinct` slice to `foundNull`
and refactors all the logic that touches it accordingly. Before the
previous commit this slice served two purposes:
- to indicate that a tuple is "distinct" because it doesn't have a hash
match, this was used by both the hash aggregator and the unordered
distinct
- to indicate that a tuple is "distinct" because it has a NULL value in
the key column and NULL values are considered different, this was used
by the unordered distinct.

The first usage was removed in the previous commit, so this commit makes
it clear the purpose of the slice to be purely for the second usage.

Release note: None

yuzefovich committed:
colexechash: refactor the hash aggregator and unordered distinct a bit

This commit refactors the hash table code when it is used by the hash
aggregator or the unordered distinct a little. In particular, now all
tuples that are known to be distinct before probing (when the tuple
doesn't have a hash match) are omitted from probing entirely. This seems
like a reasonable thing to do, but the rationale for why I wanted to do
this is to reduce the number of places where we use
`hashTableProbeBuffer.distinct` slice. Before this commit we would mark
all tuples with no hash matches as "distinct" in that slice, but for the
follow-up work on introducing the vectorized hash group joiner those
accesses of the `distinct` slice are annoying to deal with and
unnecessary.

This required templating `FindBuckets` method on a boolean that controls
how such "no hash match" tuple is handled (in some cases we want to mark
its HeadID to point to itself, but in one case when used by the hash
aggregator we want to mark its HeadID to be zero).

Additionally, in order to improve the chance of hash collisions in the
unit tests, this commit plumbs the testing knobs to reduce the number of
buckets used by the hash table.

Release note: None

yuzefovich committed:
colexechash: refactor finding buckets a bit

This commit refactors `FindBuckets` method to optimize it when probing
is done against itself. In such a scenario we know for sure that every
tuple has at least one match (the tuple itself), so `toCheckID` can
never become zero for each tuple. Additionally, when probing against
itself we see that we're about to check the `toCheckID` such that we'd
compare against the tuple itself, we already know that the match will
succeed and we can skip checking altogether.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 4 files at r1, 12 of 12 files at r4, 12 of 12 files at r5, 4 of 4 files at r6, 7 of 7 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101204#-:-NSx3CnT10VnyIlR9zYL:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/colexec/hash_aggregator.go` line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/101204#-NSwwcy483LiD84JEhri:-NSwwcy483LiD84JEhrj:bekrc7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fd4db0ca6d1c8dd600b80b090c088b2ef03f587/pkg/sql/colexec/hash_aggregator.go#L211)):*
> ```Go
> 	// TPCH queries using tpchvec/bench.
> 	hashTableNumBuckets := uint64(256)
> 	if args.TestingKnobs.HashTableNumBuckets != 0 {
> ```

[nit] Why not just set the field directly like for the unordered distinct test?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101204#-:-NT-ioX96MPv5NUAGW2D:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/colexec/hash_aggregator.go` line 211 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/101204#-NSwwcy483LiD84JEhri:-NT-i4xx996jOeB02LWF:bn5aprx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4fd4db0ca6d1c8dd600b80b090c088b2ef03f587/pkg/sql/colexec/hash_aggregator.go#L211)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Why not just set the field directly like for the unordered distinct test?
</blockquote></details>

Most of our tests run for both ordered and hash aggregators, so to modify this knob directly we'd need to do something like
```
op := agg.new(args)
if ha, ok := op.(*hashAggregator); ok {
  ha.hashTableNumBuckets = 7
}
```
Adjusting the args struct seem cleaner - the unordered distinct doesn't have the args struct and we don't have ambiguity on the type of the distinct operator.


<!-- Sent from Reviewable.io -->
